{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# All pairwise associations between Gene Expression and MicroRNA data in TCGA\n",
    "Check out more notebooks at our ['Regulome Explorer Repository'](https://github.com/isb-cgc/Community-Notebooks/tree/master/RegulomeExplorer)!\n",
    "\n",
    "\n",
    "```\n",
    "Title:   All pairwise associations between Gene Expression and MicroRNA data in TCGA\n",
    "Author:  Boris Aguilar\n",
    "Created: 03-11-2020\n",
    "Purpose: Demonstrate how to use the Regulome explorer functionality to find significant associations from Gene expression and microRNA BigQuery tables\n",
    "URL:     https://github.com/isb-cgc/Community-Notebooks/blob/master/Correlation-GeneExpression-MicroRNA\n",
    "\n",
    "\n",
    "```\n",
    "***\n",
    "\n",
    "In this notebook, we use BigQuery to find significant correlations between all possible pairs of genes in the RNAseq data and unique identifiers in the microRNA data available in TCGA. The cohort for this analysis consists of BrCA patients that are 50 years old or younger, at the time of diagnosis.\n",
    "\n",
    "We used Bigquery to compute Pearson correlations and identify significant correlation coefficients using a BigQuery [user-defined function](https://cloud.google.com/bigquery/user-defined-functions). The necessary queries are generated using python functions implemented in the [Regulome Explorer module](https://github.com/isb-cgc/Community-Notebooks/tree/master/RegulomeExplorer/re_module). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Authentication\n",
    "The first step is to authorize access to BigQuery and the Google Cloud. For more information see ['Quick Start Guide to ISB-CGC'](https://isb-cancer-genomics-cloud.readthedocs.io/en/latest/sections/HowToGetStartedonISB-CGC.html) and alternative authentication methods can be found [here](https://googleapis.github.io/google-cloud-python/latest/core/auth.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "from google.cloud import bigquery\n",
    "from scipy import stats\n",
    "import re_module.bq_functions as regulome\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parameters\n",
    "In this experiment we will use Gene Expression data and microRNA data available in ISB-CGC. The information about the Bigquery tables and their columns are stored in the Features dictionary. The significance level and the minimum number of patients for the correlation analysis are free parameteres that can be changed.\n",
    "\n",
    "If you want to test this notebook checking all possible pairs please replace \"labels1 = ...\" with \"labels1 = ' IS NOT NULL '\" but know the computations will take several minutes to complete, we recommend to generate the query and copy it in the [console](https://console.cloud.google.com/bigquery)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# information from Bigquery tables\n",
    "Features = {     'Gene Expression' : { 'table'  : 'isb-cgc.TCGA_hg38_data_v0.RNAseq_Gene_Expression',\n",
    "                                       'symbol' : 'gene_name',\n",
    "                                       'study'  : 'project_short_name',\n",
    "                                       'data'   : 'AVG( LOG10( HTSeq__Counts + 1 ) ) ',\n",
    "                                       'rnkdata': 'data',\n",
    "                                       'avgdat' : 'avgdata',  \n",
    "                                   'patientcode': 'case_barcode',\n",
    "                                    'samplecode': 'sample_barcode',\n",
    "                                       'where'  : 'AND HTSeq__Counts IS NOT NULL',\n",
    "                                       'dattype': 'numeric' },\n",
    "             'MicroRNA Expression': {  'table'  : 'isb-cgc.TCGA_hg38_data_v0.miRNAseq_Expression',\n",
    "                                       'symbol' : 'mirna_id',\n",
    "                                        'study' : 'project_short_name',\n",
    "                                       'data'   : 'AVG( reads_per_million_miRNA_mapped )',\n",
    "                                       'rnkdata': 'data',\n",
    "                                       'avgdat' : 'avgdata',\n",
    "                                   'patientcode': 'case_barcode',\n",
    "                                    'samplecode': 'sample_barcode',\n",
    "                                       'where'  : 'AND reads_per_million_miRNA_mapped IS NOT NULL',\n",
    "                                       'dattype': 'numeric'}\n",
    "                \n",
    "               }\n",
    "\n",
    "# parameteres for Gene expression data\n",
    "feat1 = Features['Gene Expression']\n",
    "cohort1 = feat1['patientcode'] + \" IN ( SELECT case_barcode FROM cohort ) \"\n",
    "labels1 = ' =  \\'IGLVV-66\\' '  #' IS NOT NULL ' use this for all pair-wise comparison; or Change IGLVV-66 for your favorite gene \n",
    "#labels1 = ' IS NOT NULL '\n",
    "\n",
    "# parameteres for Gene expression data\n",
    "feat2 = Features['MicroRNA Expression'] \n",
    "cohort2 = feat2['patientcode'] + \" IN ( SELECT case_barcode FROM cohort ) \"\n",
    "labels2 = ' IS NOT NULL '\n",
    "\n",
    "# significance level\n",
    "significance_level = 0.01 # only 0.05, 0.01, 0.005, and 0.001 are alowed !!\n",
    "\n",
    "# minimun number of samples\n",
    "nsamples = 25"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read cohort\n",
    "The cohort for this analysis consists of BrCA patients with less than 50 years  of age at the time of diagnosis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"\"\"\n",
    "cohort AS(\n",
    "SELECT case_barcode FROM `isb-cgc.TCGA_bioclin_v0.Clinical`\n",
    "WHERE project_short_name = \"TCGA-BRCA\" AND age_at_diagnosis <=50\n",
    ")\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Queries to read tables\n",
    "The queries to retreive the necessary data, query_t1 for Gene expression and query_t2 for micro RNA, are generated using a function available in the regulome explorer module available in here https://github.com/isb-cgc/Community-Notebooks/tree/master/RegulomeExplorer/re_module.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_t1 = regulome.generic_numeric_bqtable ( 'table1' , feat1, cohort1, labels1 )\n",
    "query_t2 = regulome.generic_numeric_bqtable ( 'table2' , feat2, cohort2, labels2 ) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Queries to join tables and perform statistics\n",
    "The two tables are integrated into one table to compute correlations. The query to perform these operations is obtained by using a function implemented in the regulome explorer module.  'sql' contains the complete query to perform the correlations.  You can copy the query (which is printed ) to the Bigquery - Google cloud console to run the query and obtain the results. The computations can take 15 minutes. Alternatively you can go to \"Run the Query\" section of this notebooks. \n",
    "\n",
    "Note: `cgc-05-0042.functions.significance_level_ttest2` is a [persistent function](https://cloud.google.com/bigquery/docs/reference/standard-sql/user-defined-functions)  that estimates upper bounds of p-values ( significance levels ), so we can filter out correlations that are not statistically significant ( p-value > significance_level ). This function only work for N < 500. For N>500 we recommend the normal approximation of the t distribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CREATE TEMPORARY FUNCTION tscore_to_p(a FLOAT64, b FLOAT64, c FLOAT64)\n",
      " RETURNS FLOAT64\n",
      " LANGUAGE js AS\n",
      "\"\"\"\n",
      "  return jStat.ttest(a,b,c); //jStat.ttest( tscore, n, sides)\n",
      "\"\"\"\n",
      "OPTIONS (\n",
      " library=\"gs://ba-cgc-bigquery/jstat/dist/jstat.min.js\"\n",
      ");\n",
      "\n",
      "WITH\n",
      "cohort AS(\n",
      "SELECT case_barcode FROM `isb-cgc.TCGA_bioclin_v0.Clinical`\n",
      "WHERE project_short_name = \"TCGA-BRCA\" AND age_at_diagnosis <=50\n",
      ")\n",
      ",\n",
      "table1 AS (\n",
      "SELECT\n",
      "   symbol,\n",
      "   data AS rnkdata,\n",
      "   ParticipantBarcode\n",
      "FROM (\n",
      "   SELECT\n",
      "      gene_name AS symbol, \n",
      "      AVG( LOG10( HTSeq__Counts + 1 ) )  AS data,\n",
      "      case_barcode AS ParticipantBarcode\n",
      "   FROM `isb-cgc.TCGA_hg38_data_v0.RNAseq_Gene_Expression`\n",
      "   WHERE case_barcode IN ( SELECT case_barcode FROM cohort )     # cohort \n",
      "         AND gene_name  =  'IGLVV-66'   # labels \n",
      "         AND HTSeq__Counts IS NOT NULL  \n",
      "   GROUP BY\n",
      "      ParticipantBarcode, symbol\n",
      "   )\n",
      ")\n",
      ",\n",
      "table2 AS (\n",
      "SELECT\n",
      "   symbol,\n",
      "   data AS rnkdata,\n",
      "   ParticipantBarcode\n",
      "FROM (\n",
      "   SELECT\n",
      "      mirna_id AS symbol, \n",
      "      AVG( reads_per_million_miRNA_mapped ) AS data,\n",
      "      case_barcode AS ParticipantBarcode\n",
      "   FROM `isb-cgc.TCGA_hg38_data_v0.miRNAseq_Expression`\n",
      "   WHERE case_barcode IN ( SELECT case_barcode FROM cohort )     # cohort \n",
      "         AND mirna_id  IS NOT NULL   # labels \n",
      "         AND reads_per_million_miRNA_mapped IS NOT NULL  \n",
      "   GROUP BY\n",
      "      ParticipantBarcode, symbol\n",
      "   )\n",
      ")\n",
      ",\n",
      "summ_table AS (\n",
      "SELECT \n",
      "   n1.symbol as symbol1,\n",
      "   n2.symbol as symbol2,\n",
      "   COUNT( n1.ParticipantBarcode ) as n,\n",
      "   CORR(n1.rnkdata , n2.rnkdata) as correlation\n",
      "    \n",
      "FROM\n",
      "   table1 AS n1\n",
      "INNER JOIN\n",
      "   table2 AS n2\n",
      "ON\n",
      "   n1.ParticipantBarcode = n2.ParticipantBarcode\n",
      "   \n",
      "GROUP BY\n",
      "   symbol1, symbol2\n",
      ")\n",
      "SELECT *,\n",
      "   tscore_to_p( ABS(correlation)*SQRT( (n-2)/((1+correlation)*(1-correlation))), n-2, 2) as p\n",
      "FROM summ_table\n",
      "WHERE n > 25 AND ABS(correlation) < 0.9999 AND `cgc-05-0042.functions.significance_level_ttest2`(n-2, ABS(correlation)*SQRT( (n-2)/((1+correlation)*(1-correlation)))) <= 0.01\n",
      "ORDER BY ABS(correlation) DESC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "summ_query = regulome.get_summarized_table('Gene Expression',feat1,'MicroRNA Expression',feat2)\n",
    "f_pval = regulome.query_js_tscore_pvalue()\n",
    "\n",
    "sql = (f_pval + 'WITH' + cohort + ',' + query_t1 + ',' + query_t2 + ',' + summ_query + \"\"\"\n",
    "SELECT *,\n",
    "   tscore_to_p( ABS(correlation)*SQRT( (n-2)/((1+correlation)*(1-correlation))), n-2, 2) as p\n",
    "FROM summ_table\n",
    "WHERE n > {0} AND ABS(correlation) < 0.9999 AND `cgc-05-0042.functions.significance_level_ttest2`(n-2, ABS(correlation)*SQRT( (n-2)/((1+correlation)*(1-correlation)))) <= {1}\n",
    "ORDER BY ABS(correlation) DESC\n",
    "\"\"\".format( str(nsamples) , str(significance_level) ) )\n",
    "\n",
    "print( sql )  # This Query perform the computations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run the Query\n",
    "The following steps run BigQuery and generate a table with significant associations.\n",
    "Note: alpha is an upper bound of the significant level. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " in runQuery ... \n",
      "    this query processed 22617560568 bytes \n",
      "    Approx. elpased time : 3586 miliseconds \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol1</th>\n",
       "      <th>symbol2</th>\n",
       "      <th>n</th>\n",
       "      <th>correlation</th>\n",
       "      <th>p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-124-2</td>\n",
       "      <td>323</td>\n",
       "      <td>0.419730</td>\n",
       "      <td>3.272597e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-124-3</td>\n",
       "      <td>323</td>\n",
       "      <td>0.403092</td>\n",
       "      <td>4.787322e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-5591</td>\n",
       "      <td>323</td>\n",
       "      <td>0.381459</td>\n",
       "      <td>1.261429e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-124-1</td>\n",
       "      <td>323</td>\n",
       "      <td>0.329851</td>\n",
       "      <td>1.235962e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-3935</td>\n",
       "      <td>323</td>\n",
       "      <td>0.300824</td>\n",
       "      <td>3.520930e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-6889</td>\n",
       "      <td>323</td>\n",
       "      <td>0.263086</td>\n",
       "      <td>1.632000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-153-1</td>\n",
       "      <td>323</td>\n",
       "      <td>0.258264</td>\n",
       "      <td>2.559250e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-153-2</td>\n",
       "      <td>323</td>\n",
       "      <td>0.239804</td>\n",
       "      <td>1.320803e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-1224</td>\n",
       "      <td>323</td>\n",
       "      <td>0.231927</td>\n",
       "      <td>2.560136e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-875</td>\n",
       "      <td>323</td>\n",
       "      <td>0.214430</td>\n",
       "      <td>1.027677e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-3166</td>\n",
       "      <td>323</td>\n",
       "      <td>0.209933</td>\n",
       "      <td>1.443256e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-1257</td>\n",
       "      <td>323</td>\n",
       "      <td>0.206102</td>\n",
       "      <td>1.916816e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-204</td>\n",
       "      <td>323</td>\n",
       "      <td>0.201479</td>\n",
       "      <td>2.681140e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-647</td>\n",
       "      <td>323</td>\n",
       "      <td>0.194924</td>\n",
       "      <td>4.261010e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-6796</td>\n",
       "      <td>323</td>\n",
       "      <td>0.191556</td>\n",
       "      <td>5.375027e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-4538</td>\n",
       "      <td>323</td>\n",
       "      <td>0.190952</td>\n",
       "      <td>5.601175e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-4770</td>\n",
       "      <td>323</td>\n",
       "      <td>0.188432</td>\n",
       "      <td>6.644190e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-3199-1</td>\n",
       "      <td>323</td>\n",
       "      <td>0.180772</td>\n",
       "      <td>1.101955e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-3622b</td>\n",
       "      <td>323</td>\n",
       "      <td>0.180205</td>\n",
       "      <td>1.143165e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-5682</td>\n",
       "      <td>323</td>\n",
       "      <td>0.178039</td>\n",
       "      <td>1.313693e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-564</td>\n",
       "      <td>323</td>\n",
       "      <td>0.162455</td>\n",
       "      <td>3.414031e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-5088</td>\n",
       "      <td>323</td>\n",
       "      <td>0.157966</td>\n",
       "      <td>4.429868e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>22</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-6869</td>\n",
       "      <td>323</td>\n",
       "      <td>0.155817</td>\n",
       "      <td>5.006397e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     symbol1         symbol2    n  correlation             p\n",
       "0   IGLVV-66   hsa-mir-124-2  323     0.419730  3.272597e-15\n",
       "1   IGLVV-66   hsa-mir-124-3  323     0.403092  4.787322e-14\n",
       "2   IGLVV-66    hsa-mir-5591  323     0.381459  1.261429e-12\n",
       "3   IGLVV-66   hsa-mir-124-1  323     0.329851  1.235962e-09\n",
       "4   IGLVV-66    hsa-mir-3935  323     0.300824  3.520930e-08\n",
       "5   IGLVV-66    hsa-mir-6889  323     0.263086  1.632000e-06\n",
       "6   IGLVV-66   hsa-mir-153-1  323     0.258264  2.559250e-06\n",
       "7   IGLVV-66   hsa-mir-153-2  323     0.239804  1.320803e-05\n",
       "8   IGLVV-66    hsa-mir-1224  323     0.231927  2.560136e-05\n",
       "9   IGLVV-66     hsa-mir-875  323     0.214430  1.027677e-04\n",
       "10  IGLVV-66    hsa-mir-3166  323     0.209933  1.443256e-04\n",
       "11  IGLVV-66    hsa-mir-1257  323     0.206102  1.916816e-04\n",
       "12  IGLVV-66     hsa-mir-204  323     0.201479  2.681140e-04\n",
       "13  IGLVV-66     hsa-mir-647  323     0.194924  4.261010e-04\n",
       "14  IGLVV-66    hsa-mir-6796  323     0.191556  5.375027e-04\n",
       "15  IGLVV-66    hsa-mir-4538  323     0.190952  5.601175e-04\n",
       "16  IGLVV-66    hsa-mir-4770  323     0.188432  6.644190e-04\n",
       "17  IGLVV-66  hsa-mir-3199-1  323     0.180772  1.101955e-03\n",
       "18  IGLVV-66   hsa-mir-3622b  323     0.180205  1.143165e-03\n",
       "19  IGLVV-66    hsa-mir-5682  323     0.178039  1.313693e-03\n",
       "20  IGLVV-66     hsa-mir-564  323     0.162455  3.414031e-03\n",
       "21  IGLVV-66    hsa-mir-5088  323     0.157966  4.429868e-03\n",
       "22  IGLVV-66    hsa-mir-6869  323     0.155817  5.006397e-03"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bqclient = bigquery.Client()\n",
    "df_results = regulome.runQuery ( bqclient, sql, [], [], [], dryRun=False )\n",
    "df_results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the nonsignificant correlations ( p > significance level ) were filtered out.\n",
    "We can also use python to compute p values from correlations and number of patients:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol1</th>\n",
       "      <th>symbol2</th>\n",
       "      <th>n</th>\n",
       "      <th>correlation</th>\n",
       "      <th>p</th>\n",
       "      <th>p-python</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-124-2</td>\n",
       "      <td>323</td>\n",
       "      <td>0.419730</td>\n",
       "      <td>3.272597e-15</td>\n",
       "      <td>3.108624e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-124-3</td>\n",
       "      <td>323</td>\n",
       "      <td>0.403092</td>\n",
       "      <td>4.787322e-14</td>\n",
       "      <td>4.751755e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-5591</td>\n",
       "      <td>323</td>\n",
       "      <td>0.381459</td>\n",
       "      <td>1.261429e-12</td>\n",
       "      <td>1.253664e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-124-1</td>\n",
       "      <td>323</td>\n",
       "      <td>0.329851</td>\n",
       "      <td>1.235962e-09</td>\n",
       "      <td>1.231803e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-3935</td>\n",
       "      <td>323</td>\n",
       "      <td>0.300824</td>\n",
       "      <td>3.520930e-08</td>\n",
       "      <td>3.512751e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-6889</td>\n",
       "      <td>323</td>\n",
       "      <td>0.263086</td>\n",
       "      <td>1.632000e-06</td>\n",
       "      <td>1.629772e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-153-1</td>\n",
       "      <td>323</td>\n",
       "      <td>0.258264</td>\n",
       "      <td>2.559250e-06</td>\n",
       "      <td>2.556001e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-153-2</td>\n",
       "      <td>323</td>\n",
       "      <td>0.239804</td>\n",
       "      <td>1.320803e-05</td>\n",
       "      <td>1.319548e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-1224</td>\n",
       "      <td>323</td>\n",
       "      <td>0.231927</td>\n",
       "      <td>2.560136e-05</td>\n",
       "      <td>2.557999e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-875</td>\n",
       "      <td>323</td>\n",
       "      <td>0.214430</td>\n",
       "      <td>1.027677e-04</td>\n",
       "      <td>1.027043e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-3166</td>\n",
       "      <td>323</td>\n",
       "      <td>0.209933</td>\n",
       "      <td>1.443256e-04</td>\n",
       "      <td>1.442435e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-1257</td>\n",
       "      <td>323</td>\n",
       "      <td>0.206102</td>\n",
       "      <td>1.916816e-04</td>\n",
       "      <td>1.915800e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-204</td>\n",
       "      <td>323</td>\n",
       "      <td>0.201479</td>\n",
       "      <td>2.681140e-04</td>\n",
       "      <td>2.679835e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-647</td>\n",
       "      <td>323</td>\n",
       "      <td>0.194924</td>\n",
       "      <td>4.261010e-04</td>\n",
       "      <td>4.259180e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-6796</td>\n",
       "      <td>323</td>\n",
       "      <td>0.191556</td>\n",
       "      <td>5.375027e-04</td>\n",
       "      <td>5.372866e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-4538</td>\n",
       "      <td>323</td>\n",
       "      <td>0.190952</td>\n",
       "      <td>5.601175e-04</td>\n",
       "      <td>5.598950e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-4770</td>\n",
       "      <td>323</td>\n",
       "      <td>0.188432</td>\n",
       "      <td>6.644190e-04</td>\n",
       "      <td>6.641679e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-3199-1</td>\n",
       "      <td>323</td>\n",
       "      <td>0.180772</td>\n",
       "      <td>1.101955e-03</td>\n",
       "      <td>1.101599e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-3622b</td>\n",
       "      <td>323</td>\n",
       "      <td>0.180205</td>\n",
       "      <td>1.143165e-03</td>\n",
       "      <td>1.142800e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-5682</td>\n",
       "      <td>323</td>\n",
       "      <td>0.178039</td>\n",
       "      <td>1.313693e-03</td>\n",
       "      <td>1.313291e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-564</td>\n",
       "      <td>323</td>\n",
       "      <td>0.162455</td>\n",
       "      <td>3.414031e-03</td>\n",
       "      <td>3.413287e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-5088</td>\n",
       "      <td>323</td>\n",
       "      <td>0.157966</td>\n",
       "      <td>4.429868e-03</td>\n",
       "      <td>4.428997e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>22</td>\n",
       "      <td>IGLVV-66</td>\n",
       "      <td>hsa-mir-6869</td>\n",
       "      <td>323</td>\n",
       "      <td>0.155817</td>\n",
       "      <td>5.006397e-03</td>\n",
       "      <td>5.005460e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     symbol1         symbol2    n  correlation             p      p-python\n",
       "0   IGLVV-66   hsa-mir-124-2  323     0.419730  3.272597e-15  3.108624e-15\n",
       "1   IGLVV-66   hsa-mir-124-3  323     0.403092  4.787322e-14  4.751755e-14\n",
       "2   IGLVV-66    hsa-mir-5591  323     0.381459  1.261429e-12  1.253664e-12\n",
       "3   IGLVV-66   hsa-mir-124-1  323     0.329851  1.235962e-09  1.231803e-09\n",
       "4   IGLVV-66    hsa-mir-3935  323     0.300824  3.520930e-08  3.512751e-08\n",
       "5   IGLVV-66    hsa-mir-6889  323     0.263086  1.632000e-06  1.629772e-06\n",
       "6   IGLVV-66   hsa-mir-153-1  323     0.258264  2.559250e-06  2.556001e-06\n",
       "7   IGLVV-66   hsa-mir-153-2  323     0.239804  1.320803e-05  1.319548e-05\n",
       "8   IGLVV-66    hsa-mir-1224  323     0.231927  2.560136e-05  2.557999e-05\n",
       "9   IGLVV-66     hsa-mir-875  323     0.214430  1.027677e-04  1.027043e-04\n",
       "10  IGLVV-66    hsa-mir-3166  323     0.209933  1.443256e-04  1.442435e-04\n",
       "11  IGLVV-66    hsa-mir-1257  323     0.206102  1.916816e-04  1.915800e-04\n",
       "12  IGLVV-66     hsa-mir-204  323     0.201479  2.681140e-04  2.679835e-04\n",
       "13  IGLVV-66     hsa-mir-647  323     0.194924  4.261010e-04  4.259180e-04\n",
       "14  IGLVV-66    hsa-mir-6796  323     0.191556  5.375027e-04  5.372866e-04\n",
       "15  IGLVV-66    hsa-mir-4538  323     0.190952  5.601175e-04  5.598950e-04\n",
       "16  IGLVV-66    hsa-mir-4770  323     0.188432  6.644190e-04  6.641679e-04\n",
       "17  IGLVV-66  hsa-mir-3199-1  323     0.180772  1.101955e-03  1.101599e-03\n",
       "18  IGLVV-66   hsa-mir-3622b  323     0.180205  1.143165e-03  1.142800e-03\n",
       "19  IGLVV-66    hsa-mir-5682  323     0.178039  1.313693e-03  1.313291e-03\n",
       "20  IGLVV-66     hsa-mir-564  323     0.162455  3.414031e-03  3.413287e-03\n",
       "21  IGLVV-66    hsa-mir-5088  323     0.157966  4.429868e-03  4.428997e-03\n",
       "22  IGLVV-66    hsa-mir-6869  323     0.155817  5.006397e-03  5.005460e-03"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if not df_results.empty:\n",
    "    f = lambda n, correlation  : (1.0 - stats.t.cdf( abs(correlation) * np.sqrt( (n- 2.0) / (1.0 - correlation**2 )), n-2)) * 2.0  \n",
    "    df_results['p-python'] = df_results.apply(lambda x: f(x.n,x.correlation), axis=1)\n",
    "\n",
    "df_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
